### PR TITLE
Basic search functionality

### LINKED ISSUE #129: Feature search
I propose adding an autocomplete-style site search to the top of platatus that will let users quickly find features by doing live text search. Think something like the search box on status.modern.ie or caniuse.com.

I'm planning on building it out using https://github.com/olivernn/lunr.js (which allows for fast indexing and full-text search). I can digest status.json to build the search index on page load.

Thoughts? This is a 4th-quarter deliverable for me.

### PR DESCRIPTION
**Pull request by potch**
Basic search functionality with lunr is available on this branch. This is as-of-yet a work-in-progress and point of discussion for the overall search feature.

Breakdown of the search feature as I see it in my mind

- [x] Import `lunr` library
- [x] Basic search (features show/hide based on live typing)
- [x] Search styling (open questions about appearance)
- [x] "Clear Search" button/link
- [x] Result ranking by relevance (right now alpha order is preserved)
- [ ] Tests!

Big questions are around styling. The `#features` element is only as big as its widest visible child's intrinsic width. This means as you search the column randomly resizes. Is this intended or can we set the width on the list?

We could potentially use flexbox for the features list. This would allow for purely style-based re-ordering of results using the `order` property. Might have layout perf implications, need to investigate.

Would appreciate an initial code once-through and thoughts on the direction!

This will resolve #129.

### CODE REVIEW

**Comment by potch** [comment]
Added some draft styling!

**Comment by potch** [comment]
screen shot 2015-11-18 at 15 59 26

**Comment by digitarald** [comment]
Would it fit into the header to reduce space? This would also help to make the header more appy.

**Comment by potch** [comment]
It could fit, but would suggest more of a site-wide search than an in-page filtering.

**Comment by potch** [comment]
Added clear button and UI showing number of results. Also used `order` to re-arrange results on the page.
screen shot 2015-11-19 at 15 36 08

**Comment by digitarald** [comment]
Shouldn't searching "WebAssembly" yield results?

**Comment by potch** [comment]
@digitarald yup it should- looks like the HTML in that item's summary is messing with the search code. Will look into it.

**Comment by potch** [comment]
@digitarald did a simple regex to strip out the HTML tags from the summary. searching WebAssembly now returns the expected results.

**Comment by potch** [comment]
I'm working on tests, and then this is ready to merge! Any additional feedback?

**Comment by digitarald** [comment]
Great. We have to wrap up v1 to add v2 features, so this pull request will wait a bit to be merged. We could also start a branch for v2. Open for ideas on a branching model - @brendandahl @mykmelez ?

**Comment by wfwalker** [comment]
I would love to have a staging server or some other manifestation of platatus with search; it would be cool to show off at mozlando

**Comment by darkwing** [comment]
Wow @potch , this is great!  I'm wondering if it might be nice to use the History API to add something like "search={term}" to the URL so we can initiate a search if someone shares a link.  An enhancement, of course.

**Comment by potch** [comment]
@wfwalker http://platatus-pr-161.herokuapp.com/ theoretically is that very thing, but hasn't updated when I pushed subsequent updates. not sure how to knock that ball out of the tree.

@digitarald can I perhaps start the discussion of making search a v1 feature?

File: src/css/index.css [hunk 1]
```diff
@@ -69,18 +72,55 @@ article {
...
 }
 
+.search {
+	text-align: center;
+	font-weight: normal;
+	margin: 1em 0;
+	@media (min-width: $breakMobileLandscape) {
+		display: flex;
+		align-items: center;
+		font-size: 1.5em;
```
Comment (digitarald) [inline src/css/index.css]: Use existing variable

File: src/css/index.css [hunk 2]
```diff
@@ -69,18 +72,55 @@ article {
...
+		* + * {
+			margin-left: 1em;
```
Comment (digitarald) [inline src/css/index.css]: Use gutterX variable as basis

File: src/css/index.css [hunk 3]
```diff
@@ -69,18 +72,55 @@ article {
...
+		}
+	}
+}
+.search-input {
+	font: inherit;
+	font-weight: 300;
+	flex: 1;
+	width: 100%;
```
Comment (digitarald) [inline src/css/index.css]: Shouldn't flex 1 already give it 100% width?

File: src/css/index.css [hunk 4]
```diff
@@ -69,18 +72,55 @@ article {
...
+	padding: 0 .5em;
+	line-height: 1.7em;
```
Comment (digitarald) [inline src/css/index.css]: Use `$gutterY` variables for `line-height` and `padding` for consistent vertical rhythm

File: src/css/index.css [hunk 5]
```diff
@@ -69,18 +72,55 @@ article {
...
+}
+.results-meta {
+	display: none;
+	background: #eee;
+	padding: .5em 1em;
```
Comment (digitarald) [inline src/css/index.css]: Again, use variables

File: src/css/index.css [hunk 6]
```diff
@@ -69,18 +72,55 @@ article {
...
+
 .feature {
-	margin: 0;
-	padding: 0;
-	padding-left: $gutterX;
+	margin: 0 -1em;
+	padding: 0 1em;
 	@media (min-width: $breakMobileLandscape) {
-		padding: calc($gutterX / 2) $gutterX;
+		padding: calc($gutterX / 2) 1em;
```
Comment (digitarald) [inline src/css/index.css]: What does this fix?

File: src/js/index.js [hunk 7]
```diff
@@ -51,3 +53,81 @@ if ('serviceWorker' in navigator) {
       didUpdate();
     });
 }
+
+search.initialize(function (err, index) {
```
Comment (digitarald) [inline src/js/index.js]: nit: Can we use a promise here for consistent style?

File: src/js/index.js [hunk 8]
```diff
@@ -51,3 +53,81 @@ if ('serviceWorker' in navigator) {
...
+  if (err) {
+    console.error(err);
+    return;
+  }
+
+  var queryEl = document.querySelector('.search-input');
```
Comment (digitarald) [inline src/js/index.js]: Since we are using ES6, `const` should be used here.

File: src/js/index.js [hunk 9]
```diff
@@ -51,3 +53,81 @@ if ('serviceWorker' in navigator) {
...
+    searchTimeout = setTimeout(performSearch, 100);
+  }
+
+  queryEl.addEventListener('change', triggerSearch);
+  queryEl.addEventListener('input', triggerSearch);
+  queryEl.addEventListener('keypress', triggerSearch);
+
+  clearEl.addEventListener('click', function (e) {
+    e.preventDefault();
+    queryEl.value = '';
```
Comment (digitarald) [inline src/js/index.js]: nit: Maybe focus the element after clearing?

Comment (potch) [inline src/js/index.js]: good idea!

File: src/js/index.js (hunk 7 above)
Comment (potch) [inline src/js/index.js]: A promise is sort-of hard to use internally, but the external interface can definitely be a promise.

File: src/css/index.css (hunk 3 above)
Comment (potch) [inline src/css/index.css]: at mobile sizes, the container is no longer flexed.

File: src/css/index.css (hunk 2 above)
Comment (potch) [inline src/css/index.css]: gutterX was too much spacing

File: src/css/index.css (hunk 4 above)
Comment (potch) [inline src/css/index.css]: will play with it, but mathematically pleasing numbers don't always make the most visually pleasing result.

File: src/js/index.js (hunk 9 above)
Comment (potch) [inline src/js/index.js]: fixed in 258752c

File: src/js/index.js (hunk 8 above)
Comment (potch) [inline src/js/index.js]: done in dfc8794

File: src/js/index.js (hunk 7 above)
Comment (potch) [inline src/js/index.js]: Never mind- I re-wrote using promises in dfc8794 and it's way better :)

**Comment by potch** [comment]
@digitarald Updated code to use variables!

**Comment by digitarald** [comment]
Thanks a lot! Solid start on filtering. We should hook up analytics soon to see how people use it.

**Pull request merged**